44.md

```diff
@@ -0,0 +1,80 @@
+
+A special event with kind `44`, meaning "encrypted direct message". It is supposed to have the following attributes:
+
+**`content`** MUST be equal to the JSON structure `{ciphertext, nonce, v}`. `ciphertext` is base64-encoded, ALGORITHM encrypted string of anything a user wants to write, encrypted using a shared cipher generated by combining the recipient's public-key with the sender's private-key. `nonce` is base64-encoded algorithm nonce / iv. `version` is algorithm version.
+
+**`tags`** MUST contain an entry identifying the receiver of the message (such that relays may naturally forward this event to them), in the form `["p", "<pubkey, as a hex string>"]`.
+
+**`tags`** MAY contain an entry identifying the previous message in a conversation or a message we are explicitly replying to (such that contextual, more organized conversations may happen), in the form `["e", "<event_id>"]`.
+
+**Note**: By default in the [libsecp256k1](https://github.com/bitcoin-core/secp256k1) ECDH implementation, the secret is the SHA256 hash of the shared point (both X and Y coordinates). We are using this exact implementation. In NIP4, unhashed shared point was used.
+
+## Versioning
+
+Clients MUST throw a descriptive error if they receive NIP44 message, version of which they don't support.
+
+Currently defined encryption algorithms:
+
+- `0x00` - RESERVED
+- `0x01` - XChaCha20 + sha256(ecdh)
+
+## Security Warning
+
+This standard does not go anywhere near what is considered the state-of-the-art in encrypted communication between peers, and it leaks metadata in the events, therefore it must not be used for anything you really need to keep secret, and only with relays that use `AUTH` to restrict who can fetch your `kind:4` events.
+
+## Client Implementation Warning
+
+Clients *should not* search and replace public key or note references from the `.content`. If processed like a regular text note (where `@npub...` is replaced with `#[0]` with a `["p", "..."]` tag) the tags are leaked and the mentioned user will receive the message in their inbox.
+
+## Algorithm
+
+```js
+// npm install @noble/curves @noble/hashes @scure/base @stablelib/xchacha20
+import {secp256k1} from '@noble/curves/secp256k1'
+import {sha256} from '@noble/hashes/sha256'
+import {randomBytes} from '@noble/hashes/utils'
+import {base64} from '@scure/base'
+import {streamXOR as xchacha20} from '@stablelib/xchacha20'
+
+export function getConversationKey(privkeyA: string, pubkeyB: string) {
+  const key = secp256k1.getSharedSecret(privkeyA, '02' + pubkeyB)
+  return sha256(key.slice(1, 33))
+}
+
+export function encrypt(
+  privkey: string,
+  pubkey: string,
+  text: string,
+  ver = 1
+): string {
+  if (ver !== 1) throw new Error('NIP44: unknown encryption version')
+  let key = getConversationKey(privkey, pubkey)
```

Comment:
i see no reason not to pass in the nonce here, and include it in the hash.   marginal speed difference, and that way if a conversation key gets leaked, it doesn't unlock all messages.

Reply:
if main key gets leaked, the attacker would be able to calculate sub keys. The security is no different.

Reply:
if the post-hash aes key gets leaked, the attacker can't calculate anything if we use the nonce in the hash.

Reply:
> if the post-hash aes key gets leaked, the attacker can't calculate anything

Why would the sub-key get leaked and not the main key? 

Leaked main key would allow attacker to forge messages and read ALL messages if they access all historical events, since we don't do forward secrecy. How would it be different under the proposed scheme?

Reply:
I know we're back and forth on this.

Just trying to understand by which standards it's considered a good design to use new key for every mesage, when no ratcheting or FS is in place.

XChaCha clearly says re-using 1 key for 2^80 messages would bring nonce collision probability to 2^-32. Which is miniscule. For AES it's just 2^32 messages, 14 orders of magnitude smaller.

The design is clear. We calculate key once and then decrypt all 5000 conversation messages, using different IVs (IVs may not be public if auth is used).

The suggestion is to re-calculate key for every message. For which purpose? For FS, ratcheting or anon messages that would make sense.

Reply:
i suppose it's fine.   i'll be using ephemeral keys (nip59 style) anyway.   the only reason i can think that the aes key might leak is that the crypto.subtle library is getting called over and over with the same key

Reply:
pls resolve this convo.